src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,154 @@
+            first.previous = newNode;
+            first = newNode;
+        } else if (index == size) {
+            last.next = newNode;
+            newNode.previous = last;
+            last = newNode;
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            newNode.next = prev.next;
+            newNode.previous = prev;
+            prev.next = newNode;
+            newNode.next.previous = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T e : list) {
+            this.add(e);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        Node<T> node = getNodeByIndex(index);
+        T oldElement = node.element;
+        node.element = value;
+        return oldElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        T removedElement;
+        if (index == 0) {
+            removedElement = first.element;
+            first = first.next;
+            if (first == null) {
+                last = null;
+            } else {
+                first.previous = null;
+            }
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            removedElement = prev.next.element;
+            prev.next = prev.next.next;
+            if (prev.next == null) {
+                last = prev;
+            } else {
+                prev.next.previous = prev;
+            }
+        }
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        if (isEmpty()) {
+            return false;
+        }
+        if (Objects.equals(first.element, object)) {
+            first = first.next;
+            if (first == null) {
+                last = null;
+            } else {
+                first.previous = null;
+            }
+            size--;
+            return true;
+        }
+        Node<T> current = first;
+        while (current.next != null && !Objects.equals(current.next.element, object)) {
+            current = current.next;
+        }
+        if (current.next != null) {
+            if (current.next == last) {
+                last = current;
+            }
+            current.next.previous = current;
+            current.next = current.next.next;
+            size--;
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return first == null;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current = first;
```

Comment:
iterate from last(tail) if it's closer
![image](https://github.com/mate-academy/jv-linked-list/assets/49767049/1d34d219-2cf5-4a43-97c5-597db1465fbd)